### PR TITLE
8306623: (bf) CharBuffer::allocate throws unexpected exception type with some CharSequences

### PR DESCRIPTION
**Pull request by bplb**
When appending a `CharBuffer`, perform a specific check for buffer overflow before invoking `put()`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306623](https://bugs.openjdk.org/browse/JDK-8306623): (bf) CharBuffer::allocate throws unexpected exception type with some CharSequences

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [118076c9](https://git.openjdk.org/jdk/pull/13632/files/118076c94ae74d2a4967b8b6ec12e83365f196c6)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13632/head:pull/13632` \
`$ git checkout pull/13632`

Update a local copy of the PR: \
`$ git checkout pull/13632` \
`$ git pull https://git.openjdk.org/jdk.git pull/13632/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13632`

View PR using the GUI difftool: \
`$ git pr show -t 13632`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13632.diff">https://git.openjdk.org/jdk/pull/13632.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13632#issuecomment-1521042262)

### CODE REVIEW

**Comment by bplb** [comment]
I'm not entirely comfortable with where I've put the test in this patch.

**Comment by AlanBateman** [comment]
Can you add a test to test/jdk/java/nio/Buffer/Basic-X.java.template, I think this regression was missed because this tests didn't check this case.

File: src/java.base/share/classes/java/nio/X-Buffer.java.template [hunk 1]
```diff
@@ -2045,8 +2045,17 @@ public abstract sealed class $Type$Buffer
      */
     public $Type$Buffer append(CharSequence csq, int start, int end) {
         if (csq instanceof CharBuffer cb) {
-            int pos = position();
+            //
+            // pre-emptively check for overflow as put(int,CharBuffer,int,int)
+            // would throw IndexOutOfBoundsException instead
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/X-Buffer.java.template:2050]: The comment is confusing, I think you just want to say that the append method throws BufferOverflowException when there is insufficient space in the buffer.

Comment (bplb) [reply src/java.base/share/classes/java/nio/X-Buffer.java.template]: Comment updated and test code moved in 118076c.

**Comment by bplb** [comment]
> [...] I think this regression was missed because this tests didn't check this case.

It definitely was.

**Review by AlanBateman — APPROVED** [review]

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**